packages/vite/src/node/plugins/esbuild.ts

```diff
@@ -44,11 +95,43 @@ export async function transformWithEsbuild(
     loader = 'js'
   }
 
+  const tsconfigRaw: TSConfigJSON = { compilerOptions: {} }
+  if (loader === 'ts' || loader === 'tsx') {
+    const tsconfigJson = await loadTsconfigJsonForFile(filename)
+
+    // these fields would affect the compilation result
+    // https://esbuild.github.io/content-types/#tsconfig-json
+    for (const field of [
+      'jsxFactory',
+      'jsxFragmentFactory',
+      'useDefineForClassFields',
+      'importsNotUsedAsValues'
+    ]) {
+      if (
+        tsconfigJson.compilerOptions &&
+        field in tsconfigJson.compilerOptions
+      ) {
+        // @ts-ignore
+        tsconfigRaw.compilerOptions[field] = tsconfigJson.compilerOptions[field]
+      }
+    }
+
+    // align with TypeScript 4.3
+    // https://github.com/microsoft/TypeScript/pull/42663
+    if (
+      tsconfigJson.compilerOptions?.target?.toLocaleLowerCase() === 'esnext'
+    ) {
+      tsconfigRaw.compilerOptions!.useDefineForClassFields =
```

Comment:
**suggestion:** The `!` should not be needed cause the if already flow analyse it, doesn't it?
Otherwise you may need to save `tsconfigJson.compilerOptions` to a local `const`

Reply:
Fixed.